src/Functions/array/arrayShuffle.cpp

```diff
@@ -0,0 +1,218 @@
+        return val->getUInt(0);
+    }();
+    pcg64_fast rng(seed);
+
+    size_t limit = [&]() -> size_t
+    {
+        if constexpr (Traits::has_limit)
+        {
+            if (arguments.size() > 1)
+            {
+                const auto * val = arguments[1].column.get();
+                return val->getUInt(0);
+            }
+        }
+        return 0;
+    }();
+
+    return executeGeneric(*array, rng, limit);
+}
+
+template <typename Traits>
+ColumnPtr FunctionArrayShuffleImpl<Traits>::executeGeneric(const ColumnArray & array, pcg64_fast & rng, size_t limit [[maybe_unused]])
+{
+    const ColumnArray::Offsets & offsets = array.getOffsets();
+
+    size_t size = offsets.size();
+    size_t nested_size = array.getData().size();
+    IColumn::Permutation permutation(nested_size);
+    std::iota(std::begin(permutation), std::end(permutation), 0);
+
+    ColumnArray::Offset current_offset = 0;
+    for (size_t i = 0; i < size; ++i)
+    {
+        auto next_offset = offsets[i];
+        if constexpr (Traits::has_limit)
+        {
+            if (limit)
+            {
+                const auto effective_limit = std::min<size_t>(limit, next_offset - current_offset);
+                partial_shuffle(&permutation[current_offset], &permutation[next_offset], effective_limit, rng);
+            }
+            else
+                shuffle(&permutation[current_offset], &permutation[next_offset], rng);
+        }
+        else
+            shuffle(&permutation[current_offset], &permutation[next_offset], rng);
+        current_offset = next_offset;
+    }
+    return ColumnArray::create(array.getData().permute(permutation, 0), array.getOffsetsPtr());
+}
+
+REGISTER_FUNCTION(ArrayShuffle)
+{
+    factory.registerFunction<FunctionArrayShuffleImpl<FunctionArrayShuffleTraits>>(
+        {
+            R"(
+Returns an array of the same size as the original array containing the elements in shuffled order.
+Elements are being reordered in such a way that each possible permutation of those elements has equal probability of appearance.
+
+If no seed is provided a random one will be used:
+[example:random_seed]
+
+It is possible to override the seed to produce stable results:
+[example:explicit_seed]
+)",
+         Documentation::Examples{
+                {"random_seed", "SELECT arrayShuffle([1, 2, 3, 4])"},
+                {"explicit_seed", "SELECT arrayShuffle([1, 2, 3, 4], 41)"}},
+            Documentation::Categories{"Array"}
+        },
+        FunctionFactory::CaseInsensitive);
+    factory.registerFunction<FunctionArrayShuffleImpl<FunctionArrayPartialShuffleTraits>>(
+        {
+            R"(
+Returns an array of the same size as the original array where elements in range [0..limit) are a random
+subset of the original array. Remaining [limit..n) shall contain the elements not in [0..limit) range in undefined order.
+Value of limit shall be in range [0..n]. Values outside of that range are equivalent to performing full arrayShuffle:
```

Comment:
It's ok  eitgher to keed doc only in cpp and have duplicating md. One in cpp is source of truth.

---

What the purpose of this function? Is it similar to `arrayConcat(arrayShuffle(arraySlice(a, 1, limit)), arraySlice(a, limit+1))` but with no guaranteed order for tail?

---

Also array indexing in ClickHouse are 1-based, so description can be adjusted

Reply:
As for the shuffling logic - we do the following.


For the full shuffle:
```
for i∈[0, n-2):
	j ← random from ∈ [i, n)
	swap arr[i] ↔ arr[j]
```

And for the partial shuffle we do almost exactly the same:
```
for i∈[0, limit):				// Swap only "limit" number of times
	j ← random from ∈ [i, n)		// But swap using the whole array contents
	swap arr[i] ↔ arr[j]
```
Partial shuffle is like [the one in D](https://dlang.org/library/std/random/partial_shuffle.html). We get `limit` elements ~drawn from the whole array.

This is different from the following - here we permute only a small range of elements with themselves:
```sql
SELECT
    groupArray(x) AS arr,
    3 AS lim,
    arrayConcat(arrayShuffle(arraySlice(arr, 1, lim), 41), arraySlice(arr, lim + 1)) AS subsetShuffle,
    arrayPartialShuffle(arr, lim, 41) AS partialShuffle
FROM
(
    SELECT number AS x
    FROM system.numbers
    LIMIT 10
)

Query id: 79b13717-3358-479b-96c5-a6d2c02463f0

┌─arr───────────────────┬─lim─┬─subsetShuffle─────────┬─partialShuffle────────┐
│ [0,1,2,3,4,5,6,7,8,9] │   3 │ [2,1,0,3,4,5,6,7,8,9] │ [4,8,6,3,0,5,2,7,1,9] │
└───────────────────────┴─────┴───────────────────────┴───────────────────────┘
```

---

I'll adjust the doc about indexing.

Reply:
Got it, thanks for the explanation